DistroLauncher/OOBE.cpp

```diff
@@ -161,23 +164,68 @@ namespace DistributionInfo {
 				return S_OK;
 			}
 
+			// Before relaunching, give WSL some time to make sure 
+			// distro is stopped.
+			bool stopSuccess = EnsureStopped(30u);
+			if (!stopSuccess) {
+				// We could try again, but who knows why
+				// we failed to stop the distro in the first time.
+				return ERROR_FAIL_SHUTDOWN;
+			}
+
 			// We could, but may not want to just `wsl -d Distro`.
 			// We can explore running our launcher in the future.
 			TCHAR launcherName[MAX_PATH];
 			DWORD fnLength = GetModuleFileName(NULL, launcherName, MAX_PATH);
 			if (fnLength == 0) {
 				return HRESULT_FROM_WIN32(GetLastError());
 			}
-			// Just to ensure we respect the returned buffer size:
-			const std::wstring_view launcher{ launcherName, fnLength };
-			cmdResult = _wsystem(launcher.data());
+
+			cmdResult = _wsystem(launcherName);
 			if (cmdResult != 0) {
 				return ERROR_FAIL_RESTART;
 			}
 
 			return S_OK;
 		} // HRESULT OOBEStatusHandling(std::wstring_view status).
 
+		// Polls WSL to ensure the distro is actually stopped.
+		bool EnsureStopped(unsigned int timeoutSeconds) {
+			const unsigned int magicNoOfConfirmations = 7;
+			unsigned int confirmStopped = 0;
+			bool distroIsRunning = true;
+			for (unsigned int i=0; i<timeoutSeconds; ++i) {
+				auto runner = Helpers::ProcessRunner(L"wsl -l --quiet --running");
+				auto exitCode = runner.run();
+				if (exitCode != 0L) {
+					// I'm sure we will want to customize those messages in the short future.
+					Helpers::PrintErrorMessage(ERROR_FAIL_SHUTDOWN);
+					return false;
+				}
+
+				// Every time we don't find the DistributionInfo::Name in process's stdout
+				// we increment the confirmStopped counter. When it hits the 
+				// magicNoOfConfirmations we are sure the distro is really stopped.
```

Comment:
The comment explains the code itself, but not the why we are trying multiple times to ensure that the distro is not in the output list. Is there a race somewhere making it not appearing and then appearing again? (or is it the case of just starting another instance at the same time?)

Reply:
No race in the launcher itself, as long as we confirm shutdown before relaunching the distro again
My reasoning was that the environment could cause some race, like a script user could run launching one single command inside that distro right after the shutdown command. Odd case. 
I tried to put a PoC in place to restart the distro subsequently to make my point here, but couldn't. I might be overreacting 😅 

I'll take that confirmation out and stop the loop at the first moment the distro name doesn't appear in the stdout.

Reply:
Indeed, this is what I think you were trying to get at. But even with that, the race still exists: if have the 7 try, right when you try to relaunch it, a script tried to relaunch it, you still had the race.

However, whatever the way the machine was relaunched after being stopped, we don’t really care: it was stopped at some point (not appearing empty value) and new wsl.conf values are now available:
- in the script which was optionally started
- in our "reboot" instance process

Does it make sense?
(and then, less complexity :p)